File: 4.median-of-two-sorted-arrays.java
/*
 * @lc app=leetcode id=4 lang=java
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start
// // need O(log(m+n))
// class Solution {
//     public double findMedianSortedArrays(int[] nums1, int[] nums2) {
//         // aa a aaaaaa
//         // bb b bb
//         // a + b / 2 in 
//         // if a > b 
//         // search()

//     }
// }

// shorterLeft <= longerRight && longerLeft <= shorterRight
// so we can make sure we binary search two arrays
// https://leetcode.com/problems/median-of-two-sorted-arrays/discuss/1701107/Java-Binary-Search-Explained
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if (nums1 == null && nums2 == null)
            return 0;
        
        // one of the array is empty
        if (nums1 == null) {
            int n = nums2.length;
            // 7 number
            // 0 1 2 3 4 5 6
            // 3 * 0.5 + 3 * 0.5
            return nums2[(n - 1) / 2] * 0.5 + nums2[n / 2] * 0.5;
            // 6 number 
            // 0 1 2 3 4 5
            // 2 * 0.5 + 3 * 0.5
        }

        if (nums2 == null) {
            int n = nums1.length;
            return nums1[(n - 1) / 2] * 0.5 + nums1[n / 2] * 0.5;
        }

        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        // choose shorter to binary search
        int n = nums1.length;
        int m = nums2.length;
        int left = 0;
        int right = n;

        // core code - very important
        while (left < right) {
            // i = n / 2
            // j = (n + m) / 2 - n / 2
            int i = (left + right) / 2;
            // i will increase, j will decrease if it's bigger
            // vice versa
            int j = (n + m) / 2 - i;
            System.out.println(i + " " + j);

            // binary search array 1 until left + left is medium
            if (nums1[i] < nums2[j - 1]) {
                left = i + 1;
            } else {
                right = i;
            }
        }

        // after this, we already filter out four part
        int first = left;
        int second = (n + m) / 2 - left;

        int shorterLeft = first == 0 ? Integer.MIN_VALUE : nums1[first - 1];
        int shorterRight = first == n ? Integer.MAX_VALUE : nums1[first];

        int longerLeft = second == 0 ? Integer.MIN_VALUE : nums2[second - 1];
        int longerRight = second == m ? Integer.MAX_VALUE : nums2[second];

        if ((n + m) % 2 == 1) {
            return Math.min(shorterRight, longerRight);
        } else {
            return Math.max(shorterLeft, longerLeft) * 0.5 + Math.min(shorterRight, longerRight) * 0.5;
        }
    }
}

// O(n+m)
// class Solution {
// public double findMedianSortedArrays(int[] nums1, int[] nums2) {
// int index1 = 0, index2 = 0, med1 = 0, med2 = 0;
// int length = nums1.length + nums2.length;

// for (int i = 0; i <= (length) / 2; i++) {
// med1 = med2;
// if (index1 == nums1.length) {
// med2 = nums2[index2];
// index2++;
// } else if (index2 == nums2.length) {
// med2 = nums1[index1];
// index1++;
// } else if (nums1[index1] < nums2[index2]) {
// med2 = nums1[index1];
// index1++;
// } else {
// med2 = nums2[index2];
// index2++;
// }
// }

// if ((length) % 2 == 0) {
// return (float) (med1 + med2) / 2;
// } else {
// return med2;
// }
// }
// }
// @lc code=end

File: cpp/4.median-of-two-sorted-arrays.cpp
/*
 * @lc app=leetcode id=4 lang=cpp
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start

#include <climits>
#include <vector>
using namespace std;

class Solution {
   public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        // Idea: We are given two sorted arrays, we will need to somehow merge them together and
        // take the median to return Idea1: the time complexity of merge two sorted arrays is
        // O(m+n), but we are required to do it in O(log(m+n)), so need another way Idea2: use the
        // characteristic of the median and binary search

        int n1 = nums1.size(), n2 = nums2.size();

        // Ensure nums1 is the smaller array for simplicity
        if (n1 > n2) return findMedianSortedArrays(nums2, nums1);

        int n = n1 + n2;               // size
        int left = (n1 + n2 + 1) / 2;  // Calculate the left partition size
        int low = 0, high = n1;        // low is 0, high is size of longer array

        // loop until binary search is complete
        while (low <= high) {
            int mid1 = (low + high) >> 1;  // Calculate mid index for nums1
            // Note: same as / 2 but faster
            int mid2 = left - mid1;  // Calculate mid index for nums2
            // mid 2 = (n1 + n2 + 1) / 2 - (low + high) / 2 = (n1 + n2 + 1) / 2 - (n1) / 2 = (n2 +
            // 1) / 2 = half of n2 in first loop

            int l1 = INT_MIN, l2 = INT_MIN;
            int r1 = INT_MAX, r2 = INT_MAX;

            // Idea: treat the two arrays as one array, partition it to two parts to ensure every
            // thing on the left is smaller than everything on the right since mid1 and mid2 are
            // related, we can always ensure that they sum up to half of the total size when we exit
            // the loop, we find our target partition Determine values of l1, l2, r1, and r2

            if (mid1 < n1) r1 = nums1[mid1];
            if (mid2 < n2) r2 = nums2[mid2];
            if (mid1 - 1 >= 0) l1 = nums1[mid1 - 1];
            if (mid2 - 1 >= 0) l2 = nums2[mid2 - 1];
            // set up the values of l1, l2, r1, r2, if mid1 or mid2 is invalid, make them remain min/max

            if (l1 <= r2 && l2 <= r1) {
                // The partition is correct, we found the median
                // because we partitioned both arrays to two parts, all left is smaller than all
                // right, and size is half, so we can get the median
                if (n % 2 == 1)
                    return max(l1, l2);
                else
                    return ((double)(max(l1, l2) + min(r1, r2))) / 2.0;
            } else if (l1 > r2) {
                // Move towards the left side of nums1
                // because we exactly partition both array, if first array is too big, we need to
                // shrink it
                high = mid1 - 1;
            } else {
                // l2 > r1, means nums2 is too big, move low to mid makes mid1 bigger and mid2
                // smaller Move towards the right side of nums1
                low = mid1 + 1;
            }
        }

        return 0;  // If the code reaches here, the input arrays were not sorted.
    }
};
// @lc code=end
